### ISSUE TITLE
YOLO v8 multilabel and csv format

### ISSUE THREAD

**Issue by clindhorst2** [ultralytics/ultralytics#5799]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I read in the YOLO v5 github that yolo formats need to be like this for multilabel classification: [https://github.com/ultralytics/yolov5/issues/308](url)

class	xcenter	ycenter	width	height	pic name
0	0.610156	0.5	0.779688	0.953125	A.001.01.1201.2022.07.15.png
2	0.610156	0.5	0.779688	0.953125	A.001.01.1201.2022.07.15.png

my classes look like:
YOLO ID	label
0	developing
1	harvestable
2	414

It also states that you will have multiple boxes one for each label like this:
![yolocatdog](https://github.com/ultralytics/ultralytics/assets/25907093/556da1eb-1798-4bef-9132-45ed34943a9a)

Did the issue of two bounding boxes get resolved to one?

Does YOLO v8 only support .txt files or does .csv files work as well?

**Comment by glenn-jocher** [comment]
@clindhorst2 hello,

Yes, the multilabel classification with YOLOv8 expects the data to be formatted as you mentioned. Each line of your annotation file should follow this format: 

class xcenter ycenter width height

And each image can have multiple lines for multilabel classification. Each line corresponds to a different object or class that is present in the same image.

Regarding your question about multiple bounding boxes, YOLOv8 creates multiple bounding boxes during its detection process by design. The model will naturally generate a bounding box for each identified class in an image. There has been no change in this regard in YOLOv8 compared to previous versions.

Finally, for your annotation files, YOLOv8 supports .txt format, not .csv. This decision to use .txt files is primarily to maintain consistency and simplicity in data formats across different platforms. Remember that each image should have a corresponding .txt file with the same name, containing its annotations.

I hope this helps clarify your questions about multilabel classification and data formats in YOLOv8! Let me know if there is anything unclear or you have more questions.

**Comment by clindhorst2** [comment]
@glenn-jocher So, the architecture of the network does not allow for multilabels in the same bounding box? Would it be possible to take an object detection bounding box, crop out the object detection box image, run a classification neural network (a network trained separately), and concatenate the output of YOLOv8 and the classification neural net for a single output? I'm assuming that this would be an extremely large network that could not be real-time detection. Or would you think that would be too massive to work?

**Comment by glenn-jocher** [comment]
@clindhorst2, regarding your query, indeed YOLOv8 assigns a single class per bounding box detected in its standard operation. However, your idea of creating a hybrid approach is theoretically possible but would come with practical challenges.

The approach you proposed, of taking an object detection bounding box, cropping the detected object, and then running a separate classifier on the cropped image could certainly be conducted as a two-step process. The first step would be object detection using YOLOv8 and the second step would be running a classification on the objects detected.

However, as you correctly pointed out, this would significantly increase the computational complexity and probably reduce the real-time detection performance. It would be like running two different models sequentially - the detection model and then the classification model - for each frame in a video, for example. It would also possibly increase the difficulty in terms of implementation and maintenance, due to having two separate models to train and manage.

Remember that YOLOv8 is already finely tuned for detection performance and accuracy, so adding another layer to the pipeline may lead to diminishing returns in practical use-cases. That being said, if multilabel detection within a single bounding box is an absolute necessity for your specific application, you may find experimental approaches like this worth exploring.

It's always insightful to discuss innovative ways to adapt and apply these models, and we encourage you to explore the best solutions for your specific use-case. Thank you for bringing up this interesting concept!

**Comment by clindhorst2** [comment]
@glenn-jocher according to the docs, I would need a class file for every regression label. Is that correct or can I put all regression labels in one file because they don't really have a class,

**Comment by glenn-jocher** [comment]
@clindhorst2 hello, and thanks for your question!

The format you decide to use depends on how you want your network to perform the detection.

If each regression label is associated with a distinctly different object class in your images, then you should have a separate class for each regression label. This will help the network to discriminate between these different objects.

If, instead, all the regression labels are associated with one kind of object or feature in your images and they are just different measurements or characteristics of the same object, then it would likely be best to have a single class for all regression labels. Here the regression labels represent the same kind of object, and you're trying to distinguish these objects based on different features or measurements related to the same object.

Please note that in either case, the labels need to be correctly assigned and consistent for the YOLOv8 network to learn the object characteristics effectively.

I hope this clears up your question! If you have any more questions, feel free to ask.

**Comment by glenn-jocher** [comment]
@clindhorst2, for YOLOv8, the location of label data relative to your images is crucial because the model training scripts expect a specific structure to locate these files. Here’s the general guideline for structuring your data:

1. **Dataset Directory**: Your dataset should have a directory where all the data is stored. This directory often includes subdirectories for images and labels.
   
2. **Images and Labels Subdirectories**: Within your dataset directory, you should have two subdirectories, one for images (e.g., 'images') and one for labels (e.g., 'labels'). The images subdirectory will contain all of your image files, and the labels subdirectory will contain all the corresponding label files.

3. **Matching Filenames**: Each image file should have a corresponding label file with the same name, but a different extension (images usually are `.jpg` or `.png`, while labels are `.txt`). For example, if you have an image named `image1.jpg`, the corresponding label file should be `image1.txt`.

4. **Label File Format**: Inside the label `.txt` files, each line should contain the class id and bounding box coordinates in the format:

   ```
   class x_center y_center width height
   ```

You will also need to define a `.yaml` file that specifies the paths of the train, validation, and possibly test sets, along with the class names. This `.yaml` file should be utilized when you run the training command, enabling the network to locate and use both your images and labels correctly.

If your labels are located in a different folder that is not in the expected structure relative to your images, you will need to move them into a ‘labels’ subdirectory within your main dataset directory, ensuring that the paths are correct in your dataset `.yaml` file.

Once everything is structured correctly, you can use the appropriate training command to begin training your YOLOv8 model on your dataset with the labels properly linked.
